File: Problems/No.399_Evaluate_Division/evaluate_division.cpp
class Solution {
public:

    // find the answer for each equation using DFS
    double DFS(double val, std::string node, std::string dest, std::set<string> &visited,
               std::unordered_map<std::string, std::unordered_map<std::string, double>> &Graph) {

        // check if visit the destination
        if ((node == dest) && (Graph.find(node) != Graph.end())) return val;
        else {

            double temp;
            for (auto &[next_node, ans]: Graph[node]) 
                if (visited.count(next_node) == 0) {

                    visited.insert(next_node);
                    temp = DFS(val*ans, next_node, dest, visited, Graph);
                    if (temp != -1.0) return temp;

                }
            
            return -1.0;

        }

    }

    vector<double> calcEquation(vector<vector<string>>& equations, vector<double>& values, vector<vector<string>>& queries) {
        
        // parameters
        std::vector<double> Equations;

        // start computing
        // build Equation Graph
        std::unordered_map<std::string, std::unordered_map<std::string, double>> Graph;
        for (int i=0; i<equations.size(); i++) {

            Graph[equations[i][0]][equations[i][1]] = values[i];
            Graph[equations[i][1]][equations[i][0]] = 1/values[i];

        }

        // start compute for each equation using DFS
        std::set<std::string> visited;
        for (auto &equation: queries) {

            visited.clear();
            Equations.push_back(DFS(1, equation[0], equation[1], visited, Graph));

        }

        return Equations;

    }
};

File: Problems/No.399_Evaluate_Division/evaluate_division.py
class Solution(object):

    # find the answer for each equation using DFS
    def DFS(self, val, node, dest, visited, Graph):
        # check if node, dest is in Graph
        if node not in Graph or dest not in Graph:
            return -1.0
        
        # check if visit the destination
        if (node == dest):
            return val
        else:
            for [next_node, ans] in Graph[node].items():
                if next_node not in visited:
                    visited.append(next_node)
                    temp = self.DFS(val*ans, next_node, dest, visited, Graph)
                    if (temp != -1.0):
                        return temp
            return -1.0


    def calcEquation(self, equations, values, queries):
        """
        :type equations: List[List[str]]
        :type values: List[float]
        :type queries: List[List[str]]
        :rtype: List[float]
        """

        # parameters
        Equations = []

        # start computing
        # build Equation Graph
        Graph = {}
        for (node1, node2), value in zip(equations, values):
            Graph.setdefault(node1, {})[node2] = value
            Graph.setdefault(node2, {})[node1] = 1/value

        # start compute for each equation using DFS
        visited = []
        for [source, dest] in queries:
            visited = []
            Equations.append(self.DFS(1, source, dest, visited, Graph))

        return Equations
